src/drawers/lineDrawer.ts

```diff
@@ -75,4 +90,103 @@ export class Line extends Drawer {
 
     context.restore();
   }
+
+  /**
+   * Collapse line geometry
+   *
+   * Assuming that there are many points that are drawn on the same coordinate,
+   * we can save a lot of render time by just drawing one line from the min to
+   * max y-coordinate of all those points.
+   */
+  private _collapseLineGeometry(ctx: CanvasRenderingContext2D, data: any[]) {
+    const d3Line = this._d3LineFactory();
+    const xFn = d3Line.x();
+    const yFn = d3Line.y();
+
+    // move to valid start
+    let i = 0;
+    for (; i <= data.length; ++i) {
+      const d = data[i];
+      if (d != null) {
+        const x = xFn(d, i, data);
+        const y = yFn(d, i, data);
+        ctx.moveTo(x, y);
+        ++i;
+        break;
+      }
+    }
+
+    const bucket: ILineBucket = {
+      min: null,
+      max: null,
+      prevX: null,
+      prevY: null,
+    };
+    // iterate over subsequent points
+    // TODO determine if we should collapse x or y or not collapse at all.
+    // For now, we assume line charts are always a function of x
+    this._bucketByX(ctx, data, i, xFn, yFn, bucket);
+
+    // finally, draw last line segment
+    if (bucket.prevX != null) {
+      ctx.lineTo(bucket.prevX, bucket.prevY);
+    }
+  }
+
+  /**
+   * Iterates over the line points collapsing points that fall on the same
+   * floored x coordinate.
+   *
+   * Once all the points with the same x coordinate are detected, we draw a
+   * single line from the min to max y coorindate.
+   *
+   * The "entrance" and "exit" lines to/from this collapsed vertical line are
+   * also drawn. This allows lines with no collapsed segments to render
+   * correctly.
+   */
+  private _bucketByX(
+      ctx: CanvasRenderingContext2D,
+      data: any[],
+      iStart: number,
+      xFn: d3Projector,
+      yFn: d3Projector,
+      bucket: ILineBucket,
+    ) {
+
+    for (let i = iStart; i <= data.length; ++i) {
+      const d = data[i];
+      if (d == null) {
+        continue;
+      }
+      const x = Math.floor(xFn(d, i, data));
+      const y = yFn(d, i, data);
+
+      if (bucket.prevX != null) {
```

Comment:
I don't care much about d3.line, I mostly just want all our data transformation to be logically contained to one location (or to at least say why we have "data transformation in the Plot" as well as "data transformation in the Drawer")

Cool thanks for looking into the `_filterDownsampling` method. If no one's even using it maybe we can just remove/deprecate it?